tests/core_test.py

```diff
@@ -515,5 +516,84 @@ def test_typecompat_named_shape(self):
     self.assertFalse(core.typecompat(aval1, aval3))
 
 
+class DynamicShapesTest(jtu.JaxTestCase):
```

Comment:
I suggest that we move this into djax_test?

Reply:
Hm we could do that, though djax_test.py is meant to test the code in jax/experimental/djax.py. Using core_test.py seemed like the right place for testing pre-existing core functions, namely `pe.trace_to_jaxpr_dynamic`. (The "dynamic" in the name is confusing but it was introduced in #3370, in the sense of 'dynamic scope rather than data dependence'.)

Reply:
I should add: so I lean towards keeping it in core_test.py, or else adding a new test file.

Reply:
Sounds fine.

Eventually we will have more tests for dynamic shapes, so it may make sense to create a new test file.